[1.18.x]  Game Tests: Testing within Games (#434)
MinecraftForge/Documentation: docs/misc/gametest.md

Game Tests
==========

Game Tests are a way to run in-game unit tests. The system was designed to be scalable and in parallel to run large numbers of different tests efficiently. Testing object interactions and behaviors are simply a few of the many applications of this framework.

Creating a Game Test
--------------------

A standard Game Test follows three basic steps:

1. A structure, or template, is loaded holding the scene on which the interaction or behavior is tested.
1. A method conducts the logic to perform on the scene.
1. The method logic executes. If a successful state is reached, then the test succeeds. Otherwise, the test fails and the result is stored within a lectern adjacent to the scene.

As such, to create a Game Test, there must be an existing template holding the initial start state of the scene and a method which provides the logic of execution.

### The Test Method

> A Game Test method is a `Consumer<GameTestHelper>` reference, meaning it takes in a `GameTestHelper` and returns nothing. For a Game Test method to be recognized, it must have a `@GameTest` annotation:

Reviewer: As we are mentioning the annotation, we should also mention somewhere that the `@GameTest` annotation has members which can be changed to modify how the game test is ran.

```java
public class ExampleGameTests {
  @GameTest
  public static void exampleTest(GameTestHelper helper) {
    // Do stuff
  }
}
```

#### Relative Positioning

> All `GameTestHelper` methods translate relative coordinates within the structure template scene to its absolute coordinates using the structure block's current location. To allow for easy conversion between relative and absolute positioning, `GameTestHelper#absolutePos` and `GameTestHelper#relativePos` can be used respectively.

Reviewer: Perhaps it would be useful to mention the `/test pos`/`/test var` commands for obtaining the positioning of the block at the player's cursor relative to the nearest test.

  @GameTest
  public static void exampleTest(GameTestHelper helper) { /*...*/ }

  // Class name is not prepended, template name is not specified
  // Template Location at 'modid:exampletest2'
  @PrefixGameTestTemplate(false)
  @GameTest
  public static void exampleTest2(GameTestHelper helper) { /*...*/ }

  // Class name is prepended, template name is specified
  // Template Location at 'modid:examplegametests.test_template'
  @GameTest(template = "test_template")
  public static void exampleTest3(GameTestHelper helper) { /*...*/ }

  // Class name is not prepended, template name is specified
  // Template Location at 'modid:test_template2'
  @PrefixGameTestTemplate(false)
  @GameTest(template = "test_template2")
  public static void exampleTest4(GameTestHelper helper) { /*...*/ }
}
```

Running Game Tests
------------------

Game Tests can be run using the `/test` command. The `test` command is highly configurable; however, only a few are of importance to running tests:

Subcommand  | Description
:---:       | :---
`run`       | Runs the specified test: `run <test_name>`.
`runall`    | Runs all available tests.
`runthis`   | Runs the nearest test to the player within 15 blocks.

> `runthese`  | Runs the nearest test to the player with 200 blocks.

Reviewer: The description for `runthese` seems a bit incomplete and copy-pasta'd:
```suggestion
`runthese`  | Runs tests within 200 blocks of the player.
```

`runfailed` | Runs all tests that failed in the previous run.

!!! note
    Subcommands follow the test command: `/test <subcommand>`.

Buildscript Configurations
--------------------------

Game Tests provide additional configuration settings within a buildscript (the `build.gradle` file) to run and integrate into different settings.

### Enabling Other Namespaces

> If the buildscript was setup as recommended, then only Game Tests under the current mod id would be enabled. To enable other namespaces to load Game Tests from, a run configuration must set the property `forge.enabledGameTestNamespaces` to a string specifying each namespace separated by a comma. If the property is empty or not set, then all namespaces will be loaded.

Reviewer: > If the buildscript was setup as recommended,

As recommended where, exactly?

Author: On the docs page I just linked.
